### PR TITLE
Assignment 3 - Code review request

### PR DESCRIPTION
**Pull request by karineek**

### CODE REVIEW

File: CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
+package codeu_assignement_3;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import static javafx.scene.input.KeyCode.T;
```
Comment (simon-frankau) [inline CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java]: This looks like an accidental import.

File: CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+       Returns whether the given string is a prefix of at 
+       least one word in thedictionary 
+    */
+    @Override
+    public boolean isPrefix (String p)
+    {
+        if (isWord(p)) 
+        {
+            return true;
+        }
```
Comment (simon-frankau) [inline CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java]: This special case is unnecessary. The word is a prefix of itself, so it's handled by the code below.

You might claim it's an optimisation, but it means we do an extra pass through all the words in the list, and adds extra unnecessary complexity.

File: CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        final ArrayList<String> dict_temp = new ArrayList<>();
+        String p_u = p.toUpperCase();
+        m_dict.forEach(str -> {
+	    if(str.startsWith(p_u)) 
+            {
+                dict_temp.add(str);
+            }
+        });
+        
+        return (dict_temp.size() > 0);
```
Comment (simon-frankau) [inline CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java]: There's no need to find all the words that p is a prefix of. Just return true as soon as we find one.

This code also seems a lot more complicated than doing a basic for loop.

File: CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+ * Consider using mocked dictionary instead of this implementation
+ * 
+ */
+public class SimpleDictionary implements KemDictionary 
+{
+    /* Dictionary as an array */
+    private final ArrayList<String> m_dict = new ArrayList<>(); 
+    
+    
+    /* Add all in BFS order (i.e., Breadth-First Traversal of a Tree) */
```
Comment (simon-frankau) [inline CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java]: BFS? I'm not sure what tree you're referring to.

File: CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        m_dict.forEach(str -> {
+	    if(str.startsWith(p_u)) 
+            {
+                dict_temp.add(str);
+            }
+        });
+        
+        return (dict_temp.size() > 0);
+    }
+}
```
Comment (simon-frankau) [inline CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java]: I understand this is a test implementation, and efficiency is not a prime concern. Have you thought about how you would implement this if you _were_ trying to make an efficient implementation?

File: CodeU_Assignement_3/test/codeu_assignement_3/SimpleDictionaryTest.java [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+    public SimpleDictionaryTest() {
+    }
+    
+    @BeforeClass
+    public static void setUpClass() {
+    }
+    
+    @AfterClass
+    public static void tearDownClass() {
+    }
```
Comment (simon-frankau) [inline CodeU_Assignement_3/test/codeu_assignement_3/SimpleDictionaryTest.java]: Don't write unnecessary, empty boilerplate.

File: CodeU_Assignement_3/test/codeu_assignement_3/SimpleDictionaryTest.java [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+
+    /**
+     * Test of isWord method, of class SimpleDictionary.
+     */
+    @Test
+    public void testIsWordNullDict() {
+        System.out.println("isWord Null dictionary");
```
Comment (simon-frankau) [inline CodeU_Assignement_3/test/codeu_assignement_3/SimpleDictionaryTest.java]: You shouldn't need to write strings like this out. The testing infrastructure should print out what it needs to.

File: CodeU_Assignement_3/test/codeu_assignement_3/SimpleDictionaryTest.java [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+        String w = "";
+        SimpleDictionary instance = new SimpleDictionary(null);
+        boolean expResult = false;
+        boolean result = instance.isWord(w);
+        assertEquals(expResult, result);
+    }
```
Comment (simon-frankau) [inline CodeU_Assignement_3/test/codeu_assignement_3/SimpleDictionaryTest.java:27]: KemDictionary dict = new SimpleDictionary(null);
assertFalse(dict.isWord(""));

You could even inline "dict" if you wanted!

File: CodeU_Assignement_3/test/codeu_assignement_3/SimpleDictionaryTest.java [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+        String p1 = "c";
+        String p2 = "ca";
+        String[] dict = new String[1];
+        dict[0] = "cat";
+        SimpleDictionary instance = new SimpleDictionary(dict);
+        boolean expResult = true;
+        boolean result1 = instance.isPrefix(p1);
+        boolean result2 = instance.isPrefix(p2);
+        assertEquals(expResult, result1);
+        assertEquals(expResult, result2);
```
Comment (simon-frankau) [inline CodeU_Assignement_3/test/codeu_assignement_3/SimpleDictionaryTest.java:124]: You could also test that "" and "cat" are prefixes of "cat" - this allows your tests to act as documentation of the corner cases.

File: CodeU_Assignement_3/test/codeu_assignement_3/SimpleDictionaryTest.java [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+        String p1 = "CA";
+        String p2 = "CAR";
+        String[] dict = {"CAR", "CARD", "CART", "CAT"};
+        SimpleDictionary instance = new SimpleDictionary(dict);
+        boolean expResult = true;
+        boolean result1 = instance.isPrefix(p1);
+        boolean result2 = instance.isPrefix(p2);
+        assertEquals(expResult, result1);
+        assertEquals(expResult, result2);
+    }
```
Comment (simon-frankau) [inline CodeU_Assignement_3/test/codeu_assignement_3/SimpleDictionaryTest.java:183]: Ah, and now we _do_ test that CAR is a prefix of CAR. It's not clear why the tests are split up like this, and because they're very verbose, it's difficult to quickly read through them and see what's going on. I suggest taking a look at some of the tests others have written.

File: CodeU_Assignement_3/test/codeu_assignement_3/SimpleDictionaryTest.java [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+    
+    @AfterClass
+    public static void tearDownClass() {
+    }
+
+    /**
+     * Test of isWord method, of class SimpleDictionary.
+     */
+    @Test
```
Comment (simon-frankau) [inline CodeU_Assignement_3/test/codeu_assignement_3/SimpleDictionaryTest.java]: These comments aren't terribly useful. The general idea with unit tests is to provide them with names that are, in themselves, clear and informative, and then you can drop the associated comments.

File: CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java [hunk 12]
```diff
@@ -1,16 +1,138 @@
...
+    Input: the number of rows, number of columns, 
+           a 2-dimensional array of characters (of the native char data type)
+           a dictionary
+    
+    Process: 
     
-    public KemWordSearch (KemDictionary dictionary) 
+    Output: return the set of all-words found
+    */
+    String[] wordSearch(int row, int col, char[][] grid, KemDictionary dictionary)
```
Comment (simon-frankau) [inline CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java:20]: Why take row, col as parameters? They can be inferred from the size of grid.

File: CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java [hunk 13]
```diff
@@ -1,16 +1,138 @@
...
+        }
+        if (dictionary == null)
+        {
+            return null;
+        }
+        
+        ArrayList<String> ret_temp = new ArrayList<>();
+
+        /* Search for all possible strings from the grid */
+        for (int r = 0 ; r < row; r++)
```
Comment (simon-frankau) [inline CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java]: No spaces before semicolons.

File: CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java [hunk 14]
```diff
@@ -1,16 +1,138 @@
...
+        {
+            for (int c = 0 ; c < row; c++)
+            {
+                /* Search for all the words from a specific location */
+                String[] ret_inner 
+                        = wordSearchFromPossition(row,col,grid,dictionary,r,c,"");
```
Comment (simon-frankau) [inline CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java]: Space after commas.

File: CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java [hunk 15]
```diff
@@ -1,16 +1,138 @@
...
+        {
+            return null;
+        }
+        
+        ArrayList<String> ret_temp = new ArrayList<>();
+
+        /* Search for all possible strings from the grid */
+        for (int r = 0 ; r < row; r++)
+        {
+            for (int c = 0 ; c < row; c++)
```
Comment (simon-frankau) [inline CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java]: "c < col"?

This suggests some extra tests would be a good idea.

File: CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java [hunk 16]
```diff
@@ -1,16 +1,138 @@
...
+            {
+                /* Search for all the words from a specific location */
+                String[] ret_inner 
+                        = wordSearchFromPossition(row,col,grid,dictionary,r,c,"");
+                
+                /* If there is something to add, add it */
+                if (ret_inner != null)
```
Comment (simon-frankau) [inline CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java]: It is simpler to have it always return a non-null value, and use a size zero array or other collection.

Special-casing for null is annoying and best avoided if you can.

File: CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java [hunk 17]
```diff
@@ -1,16 +1,138 @@
...
+                {
+                    ret_temp.addAll(Arrays.asList(ret_inner));
+                }
+            }
+        }
+        
+        /* Return the set from any start point */
+        return (ret_temp.isEmpty()) ? null : (ret_temp
```
Comment (simon-frankau) [inline CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java]: Why return null for an empty collection? An empty collection is easy to test for and means less special-case code needed for our caller.

Moreover, you could just use a Set to hold the words, and then you don't need to remove duplicates. This has the potential to save memory if the algorithm generates a large number of duplicates, as you will automatically de-duplicate as you go.

File: CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java [hunk 18]
```diff
@@ -1,16 +1,138 @@
...
+        ret_temp.addAll(Arrays.asList(wordSearchFromPossition
+                            (row,col,grid,dictionary,row_curr,col_curr +1,new_prefix)));
+        ret_temp.addAll(Arrays.asList(wordSearchFromPossition
+                            (row,col,grid,dictionary,row_curr -1,col_curr -1,new_prefix)));
+        ret_temp.addAll(Arrays.asList(wordSearchFromPossition
+                            (row,col,grid,dictionary,row_curr -1,col_curr +1,new_prefix)));
+        ret_temp.addAll(Arrays.asList(wordSearchFromPossition
+                            (row,col,grid,dictionary,row_curr +1,col_curr -1,new_prefix)));
+        ret_temp.addAll(Arrays.asList(wordSearchFromPossition
+                            (row,col,grid,dictionary,row_curr +1,col_curr +1,new_prefix)));
```
Comment (simon-frankau) [inline CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java]: A pair of nested for loops does this more neatly.

File: CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java [hunk 19]
```diff
@@ -1,16 +1,138 @@
...
+        
+        /* Current cell */
+        if (dictionary.isWord(new_prefix))
+        {
+            ret_temp.add(new_prefix);
+        }
+        
+        /* Mark the current cell as USED */
+        char current = grid[row_curr][col_curr];
+        grid[row_curr][col_curr] = '-';
```
Comment (simon-frankau) [inline CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java:100]: I'm not a great fan of this approach:
- It assumes that '-' is an invalid input character, which should be a documented constraint and is a maintenance overhead.
- It modifies the input data, and the caller may not expect this. I know it's changed back afterwards, but in a multithreaded environment this can be quite surprising.

I suggest keeping a separate table of booleans for "has been used or not".

File: CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java [hunk 20]
```diff
@@ -1,16 +1,138 @@
...
+                            (row,col,grid,dictionary,row_curr +1,col_curr -1,new_prefix)));
+        ret_temp.addAll(Arrays.asList(wordSearchFromPossition
+                            (row,col,grid,dictionary,row_curr +1,col_curr +1,new_prefix)));        
+        
+        /* Free the current location to be used in other search, that didn't use it yet! */
+        grid[row_curr][col_curr] = current;
+        
+        /* Return the strings from a single starting point */
+        String[] ret = ret_temp.toArray(new String[0]);
+        return ret;
```
Comment (simon-frankau) [inline CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java]: You could just write "ret_temp.toArray(new String[0]);".

However, this seems like a lot of unnecessary copying:
- You copy when converting between arrays and collections.
- At each recursive step, you copy the collected data into a new structure to return.

You could just create an "accumulator" container that you pass around and put the words in, rather than gluing all the containers together in the returning path.

File: CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java [hunk 21]
```diff
@@ -1,16 +1,138 @@
...
     }
 }
```
Comment (simon-frankau) [inline CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java]: I think you may still be on the verbose end of the commenting spectrum, but this is mostly "why?" rather than "what?", so this is good.

What do you think the algorithmic complexity of your algorithm is?

File: CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java [hunk 22]
```diff
@@ -0,0 +1,72 @@
+/*
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
```
Comment (simon-frankau) [inline CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java]: Please try to remove the auto-generated spam from your files.

File: CodeU_Assignement_3/test/codeu_assignement_3/KemWordSearchTest.java [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+        int col = 3;
+        char[][] grid = {{'A','T','I'},{'A','C','J'},{'R','D','K'}};
+        String[] dict_word = {"CAR", "CARD", "CART", "CAT","AT","FACE","COOL","OPEN","RAT","I","ARC"};
+        SimpleDictionary dict = new SimpleDictionary(dict_word);
+        KemWordSearch instance = new KemWordSearch();
+        String[] expResult = {"AT","I","ARC","CAR", "CARD", "CAT","RAT"};
+        String[] result = instance.wordSearch(row, col, grid,dict);
+        assertArrayEquals(expResult, result);
+    }
+}
```
Comment (simon-frankau) [inline CodeU_Assignement_3/test/codeu_assignement_3/KemWordSearchTest.java]: There are some good tests here but, as with the other file, it's written in an unnecessarily verbose style. Look at your colleagues' code for ideas on how to make the tests shorter.

I don't see any code that tests for repeated use of characters - you could have a bug where the same character is used twice, and i don't think these tests would cover it.

**Review by simon-frankau — APPROVED** [review]
While the code could still be briefer and neater, I think it's a good improvement on the previous assignments. Thanks.

File: CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java (hunk 22 above)
Comment (karineek) [reply CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java]: sure, np.
Is there a way to move it for good? I.e., when a new class created, it will not add it automatically?

File: CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java (hunk 1 above)
Comment (karineek) [reply CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java]: yep. Removed it.

File: CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java (hunk 4 above)
Comment (karineek) [reply CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java]: It doesn't really matter how we add the words. Removed.

File: CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java (hunk 2 above)
Comment (karineek) [reply CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java]: Performance issues. A "for loop" is more expensive than the containers methods ...

File: CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java (hunk 3 above)
Comment (karineek) [reply CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java]: OK, will do the return in the middle of the loop

File: CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java (hunk 5 above)
Comment (karineek) [reply CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java]: yes, you can build it as a sets of trees:
        c                     a 
    ca    cu            ap    ar  
cat can cur     app     are arr

File: CodeU_Assignement_3/test/codeu_assignement_3/SimpleDictionaryTest.java (hunk 11 above)
Comment (karineek) [reply CodeU_Assignement_3/test/codeu_assignement_3/SimpleDictionaryTest.java]: OK, removed

File: CodeU_Assignement_3/test/codeu_assignement_3/SimpleDictionaryTest.java (hunk 7 above)
Comment (karineek) [reply CodeU_Assignement_3/test/codeu_assignement_3/SimpleDictionaryTest.java]: So what can I do instead the description of the test? (line number of the fail?) I'm a bit confused...

File: CodeU_Assignement_3/test/codeu_assignement_3/SimpleDictionaryTest.java (hunk 9 above)
Comment (karineek) [reply CodeU_Assignement_3/test/codeu_assignement_3/SimpleDictionaryTest.java:124]: sure, will add

File: CodeU_Assignement_3/test/codeu_assignement_3/SimpleDictionaryTest.java (hunk 10 above)
Comment (karineek) [reply CodeU_Assignement_3/test/codeu_assignement_3/SimpleDictionaryTest.java:183]: size of the dictionary, that's how the tests are split. I'll add a comment

File: CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java (hunk 12 above)
Comment (karineek) [reply CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java:20]: It was the input in the interface in the question sent in slack, I just followed the interface...
 Saying this:
"Your function receives the number of rows, number of columns, a 2-dimensional array ofcharacters (of the native char data type), and the dictionary. You should return the set of allwords found."

File: CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java (hunk 13 above)
Comment (karineek) [reply CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java]: np

File: CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java (hunk 15 above)
Comment (karineek) [reply CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java]: Yes, that we can have words starting in any position. Will add. thanks!

File: CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java (hunk 14 above)
Comment (karineek) [reply CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java]: OK

File: CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java (hunk 16 above)
Comment (karineek) [reply CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java]: sure, thanks

File: CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java (hunk 17 above)
Comment (karineek) [reply CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java]: OK, fixed

File: CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java (hunk 19 above)
Comment (karineek) [reply CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java:100]: I  was thinking of it, but then you'll spend more memory... if it is ok with the memory, I can change it to it...

File: CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java (hunk 18 above)
Comment (karineek) [reply CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java]: Sure, fixed

File: CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java (hunk 21 above)
Comment (karineek) [reply CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java]: I never sure when they expect me to return a native data type or a container. Since the output suppose to be String[] I assume that's how we suppose in this assignment to  transfer the data.
Personally, I would push it to a data member of the class and return it upon request.

Anyhow,I fixed it how you suggested.

File: CodeU_Assignement_3/test/codeu_assignement_3/KemWordSearchTest.java (hunk 23 above)
Comment (karineek) [reply CodeU_Assignement_3/test/codeu_assignement_3/KemWordSearchTest.java]: sure will add some

**Comment by karineek** [comment]
Just pushed the fix for the code review, thanks!

File: CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java (hunk 22 above)
Comment (simon-frankau) [reply CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java]: It tells you in the comment how to change the templates and remove that boilerplate.

File: CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java (hunk 2 above)
Comment (simon-frankau) [reply CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java]: This is the wrong approach:
- This is supposed to be a simple and quick implementation of the Dictionary interface. Don't add unnecessary complexity.
- If you want it to be efficient, the correct solution is to make the main prefix search fast, not add a special case onto it.
- Even if you couldn't make the prefix search faster, this is an optimisation of marginal use. It accelerates the leaves of the search tree, ignoring that most isPrefix calls are not looking up complete words. Making the occasional complete word look-up faster will not stop the main prefix check loop from dominating the cost.
- The isWord check has the same algorithmic complexity as the prefix loop - m_dict is a list, so isWord is O(n) in the dictionary size anyway, just like the prefix-checking loop.

There's a popular saying "Premature optimisation is the root of all evil.". It's important to know when not to optimise, as well as when to optimise, and this is one of those cases.

File: CodeU_Assignement_3/test/codeu_assignement_3/SimpleDictionaryTest.java (hunk 7 above)
Comment (simon-frankau) [reply CodeU_Assignement_3/test/codeu_assignement_3/SimpleDictionaryTest.java]: Normally the test framework will tell you the name of the test method that failed, and you're expected to provide descriptive names for the test methods.

File: CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java (hunk 12 above)
Comment (simon-frankau) [reply CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java:20]: Ah, I missed this! In that case, I would just do a consistency check between the parameters and the shape of the grid. This _might_ be written this way because a raw array in C(++) doesn't give you the size, and the assignment is also being set in other languages.

File: CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java (hunk 19 above)
Comment (simon-frankau) [reply CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java:100]: It's not a huge amount of extra memory - doing it your way also feels like premature optimisation.

File: CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java (hunk 2 above)
Comment (karineek) [reply CodeU_Assignement_3/src/codeu_assignement_3/SimpleDictionary.java]: OK, removed.
Sure. it is only for testing

File: CodeU_Assignement_3/test/codeu_assignement_3/SimpleDictionaryTest.java (hunk 6 above)
Comment (karineek) [reply CodeU_Assignement_3/test/codeu_assignement_3/SimpleDictionaryTest.java]: sure, np

File: CodeU_Assignement_3/test/codeu_assignement_3/SimpleDictionaryTest.java (hunk 8 above)
Comment (karineek) [reply CodeU_Assignement_3/test/codeu_assignement_3/SimpleDictionaryTest.java:27]: yep, it really shorten the code!

**Comment by karineek** [comment]
finish to changes
